File: app/src/main/java/me/devsaki/hentoid/ui/CarouselDecorator.java
package me.devsaki.hentoid.ui;

import android.content.Context;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.LinearSnapHelper;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

public class CarouselDecorator {

    private final Context context;
    private final int itemLayout;
    private final CarouselAdapter adapter;
    private final LinearLayoutManager layoutManager;

    private int pageCount;
    private OnPageChangeListener onPageChangeListener;

    public CarouselDecorator(Context context, @LayoutRes int itemLayout) {
        this.context = context;
        this.itemLayout = itemLayout;
        adapter = new CarouselAdapter();
        layoutManager = new LinearLayoutManager(context);
        layoutManager.setOrientation(LinearLayoutManager.HORIZONTAL);
    }

    public void setPageCount(int pageCount) {
        this.pageCount = pageCount;
        adapter.notifyDataSetChanged();
    }

    public void setCurrentPage(int page) {
        layoutManager.scrollToPosition(page - 1);
    }

    public void setOnPageChangeListener(OnPageChangeListener onPageChangeListener) {
        this.onPageChangeListener = onPageChangeListener;
    }

    public void decorate(RecyclerView recyclerView) {
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(adapter);
        recyclerView.addOnScrollListener(new CarouselOnScrollListener());

        LinearSnapHelper snapHelper = new LinearSnapHelper();
        snapHelper.attachToRecyclerView(recyclerView);
    }

    private class CarouselOnScrollListener extends RecyclerView.OnScrollListener {

        @Override
        public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
            if (newState != RecyclerView.SCROLL_STATE_IDLE) return;

            int position = layoutManager.findFirstVisibleItemPosition();
            onPageChangeListener.onPageChange(position + 1);
        }
    }

    private class CarouselAdapter extends RecyclerView.Adapter<CarouselViewHolder> {

        @NonNull
        @Override
        public CarouselViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(context);
            View view = inflater.inflate(itemLayout, parent, false);
            return new CarouselViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull CarouselViewHolder holder, int position) {
            holder.textView.setText(String.valueOf(position + 1));
        }

        @Override
        public int getItemCount() {
            return pageCount;
        }
    }

    private static class CarouselViewHolder extends RecyclerView.ViewHolder {

        private final TextView textView;

        private CarouselViewHolder(View itemView) {
            super(itemView);
            textView = (TextView) itemView;
        }
    }

    public interface OnPageChangeListener {
        void onPageChange(int page);
    }
}

File: app/src/main/java/me/devsaki/hentoid/ui/CarouselDecorator.kt
package me.devsaki.hentoid.ui

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnTouchListener
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearSnapHelper
import androidx.recyclerview.widget.RecyclerView
import me.devsaki.hentoid.core.Consumer
import kotlin.math.max

class CarouselDecorator(context: Context, @LayoutRes private val itemLayout: Int) {
    private val adapter: CarouselAdapter
    private val layoutManager: LinearLayoutManager

    private var pageCount = 0
    private var onPageChangeListener: Consumer<Int>? = null
    private var touchListener: OnTouchListener? = null

    init {
        adapter = CarouselAdapter()
        layoutManager = LinearLayoutManager(context)
        layoutManager.orientation = LinearLayoutManager.HORIZONTAL
    }

    fun setPageCount(pageCount: Int) {
        this.pageCount = pageCount
        adapter.notifyDataSetChanged()
    }

    fun setCurrentPage(page: Int) {
        layoutManager.scrollToPosition(page - 1)
    }

    fun setOnPageChangeListener(onPageChangeListener: Consumer<Int>?) {
        this.onPageChangeListener = onPageChangeListener
    }

    fun setTouchListener(touchListener: OnTouchListener?) {
        this.touchListener = touchListener
    }

    fun decorate(recyclerView: RecyclerView) {
        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = adapter
        recyclerView.addOnScrollListener(CarouselOnScrollListener())
        val snapHelper = LinearSnapHelper()
        snapHelper.attachToRecyclerView(recyclerView)
    }

    private inner class CarouselOnScrollListener : RecyclerView.OnScrollListener() {
        override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
            if (newState != RecyclerView.SCROLL_STATE_IDLE) return
            val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()
            val lastCompletelyVisibleItemPosition =
                layoutManager.findLastCompletelyVisibleItemPosition()
            onPageChangeListener?.invoke(
                max(
                    firstVisibleItemPosition,
                    lastCompletelyVisibleItemPosition
                ) + 1
            )
        }
    }

    private inner class CarouselAdapter :
        RecyclerView.Adapter<CarouselViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CarouselViewHolder {
            val inflater = LayoutInflater.from(parent.context)
            val view: View = inflater.inflate(itemLayout, parent, false)
            view.setOnTouchListener(touchListener)
            return CarouselViewHolder(view)
        }

        override fun onBindViewHolder(holder: CarouselViewHolder, position: Int) {
            val displayPos = position + 1
            holder.textView.text = "$displayPos"
        }

        override fun getItemCount(): Int {
            return pageCount
        }
    }

    private inner class CarouselViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        val textView: TextView

        init {
            textView = itemView as TextView
        }
    }
}